instantsearch/src/main/java/com/algolia/instantsearch/ui/databinding/BindingHelper.java

```diff
@@ -1,128 +1,193 @@
+            mapAttribute(attribute, id, variant);
+        }
     }
 
-    @SuppressWarnings({"unused"}) // called via Data Binding
-    @Deprecated // Should not be used by library users
-    @BindingAdapter({"attribute", "highlighted", "highlightingColor"})
-    public static void bindHighlighted(@NonNull View view, String attribute, Boolean isHighlighted, @NonNull String colorStr) {
-        // Bind attribute, enable highlight with color
-        bindAndHighlight(view, attribute, colorStr);
+    /**
+     * Gets the current bindings.
+     *
+     * @return a SparseArray mapping each View's id to its attribute name.
+     */
+    public static HashMap<Integer, String> getBindings(String indexVariant) {
+        return bindings.get(indexVariant);
     }
 
-    @SuppressWarnings({"unused", "UnusedParameters"}) // called via Data Binding and throws
-    @Deprecated // Should not be used by library users
-    @BindingAdapter({"highlighted"})
-    public static void bindInvalid(@NonNull View view, Boolean isHighlighted) {
-        throwBindingError(view, Errors.BINDING_HIGHLIGHTED_NO_ATTR);
+    /**
+     * Associates programmatically a view with its variant, taking it from its `variant` layout attribute.
+     *
+     * @param view  any existing view.
+     * @param attrs the view's AttributeSet.
+     * @return the previous variant for this view, if any.
+     */
+    @Nullable
+    public static String setVariantForView(@NonNull View view, @NonNull AttributeSet attrs) {
+        String previousVariant;
+        final TypedArray styledAttributes = view.getContext().getTheme()
+                .obtainStyledAttributes(attrs, R.styleable.View, 0, 0);
+        try {
+            previousVariant = setVariantForView(view, styledAttributes
+                    .getString(R.styleable.View_variant));
+        } finally {
+            styledAttributes.recycle();
+        }
+
+        return previousVariant;
     }
 
-    @SuppressWarnings({"unused", "UnusedParameters"}) // called via Data Binding and throws
-    @Deprecated // Should not be used by library users
-    @BindingAdapter({"highlightingColor"})
-    public static void bindInvalid(@NonNull View view, @ColorRes int color) {
-        throwBindingError(view, Errors.BINDING_HIGHLIGHTING_NO_ATTR);
+    /**
+     * Associates programmatically a view with a variant.
+     *
+     * @param view any existing view.
+     * @return the previous variant for this view, if any.
+     */
+    @Nullable public static String setVariantForView(@NonNull View view, @Nullable String variant) {
+        String previousVariant = null;
+        String previousAttribute = null;
+        for (Map.Entry<String, HashMap<Integer, String>> entry : bindings.entrySet()) {
+            if (entry.getValue().containsKey(view.getId())) {
+                previousVariant = entry.getKey();
+                previousAttribute = entry.getValue().get(view.getId());
+                bindings.remove(previousVariant);
```

Comment:
Is this safe? Modifying the collection while iterating ? I would do a two step process. First find then remove

Reply:
I thought iterating on the entrySet was safe, but you're right: